File: algorithms/hk-it-vv/cpp/hk_it_vv.cpp
/*
 * Hopcroft-Karp Iterative Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * VV (vector-of-vectors) adjacency.
 * Old HK's lean BFS (single sLevel[] array, sentinel trick) + iterative
 * stack-based DFS with edge index array (no recursion, no rescan).
 *
 * Refactored to separate input (BipartiteGraph), output (BipartiteMatching),
 * and algorithm state (HKIState).
 */

#include <cstdio>
#include <cstdlib>
#include <cstdint>
#include <vector>
#include <algorithm>
#include <string>
#include <chrono>
#include <climits>

constexpr int32_t NIL = -1;

/* ---------- Input: BipartiteGraph ---------- */

struct BipartiteGraph {
    size_t sNumVtxs;
    size_t tNumVtxs;
    size_t numEdgs;
    std::vector<std::vector<int32_t>> sAdj;
    std::vector<std::vector<int32_t>> tAdj;
};

/* ---------- Output: BipartiteMatching ---------- */

struct BipartiteMatching {
    size_t sNumVtxs;
    size_t tNumVtxs;
    size_t numEdgs;
    std::vector<int32_t> sMate;
    std::vector<int32_t> tMate;
};

/* ---------- State: HKIState ---------- */

struct HKIState {
    std::vector<int32_t> sLevel;         // length sNumVtxs+1; sLevel[sNumVtxs] is the NIL sentinel
    std::vector<int32_t> sIdx;           // length sNumVtxs; relative offset within s's adjacency, persistent within a phase
    std::vector<int32_t> sPrcbStk;       // length sNumVtxs; DFS stack (s-vertices)
    std::vector<int32_t> tPrcbStk;       // length sNumVtxs; t-vertex chosen at each depth
    int32_t stkTop;                      // current depth of stack
};

/* ---------- BipartiteGraph construction ---------- */

BipartiteGraph buildBipartiteGraph(size_t sNumVtxs, size_t tNumVtxs,
                                   const std::vector<std::pair<int32_t,int32_t>>& edges) {
    BipartiteGraph graph;
    graph.sNumVtxs = sNumVtxs;
    graph.tNumVtxs = tNumVtxs;
    graph.sAdj.resize(sNumVtxs);
    graph.tAdj.resize(tNumVtxs);
    for (auto& e : edges) {
        int32_t s = e.first, t = e.second;
        if (s >= 0 && static_cast<size_t>(s) < sNumVtxs && t >= 0 && static_cast<size_t>(t) < tNumVtxs) {
            graph.sAdj[s].push_back(t);
            graph.tAdj[t].push_back(s);
        }
    }
    for (size_t s = 0; s < sNumVtxs; s++) {
        std::sort(graph.sAdj[s].begin(), graph.sAdj[s].end());
        graph.sAdj[s].erase(std::unique(graph.sAdj[s].begin(), graph.sAdj[s].end()), graph.sAdj[s].end());
    }
    for (size_t t = 0; t < tNumVtxs; t++) {
        std::sort(graph.tAdj[t].begin(), graph.tAdj[t].end());
        graph.tAdj[t].erase(std::unique(graph.tAdj[t].begin(), graph.tAdj[t].end()), graph.tAdj[t].end());
    }
    size_t total = 0;
    for (size_t s = 0; s < sNumVtxs; s++) total += graph.sAdj[s].size();
    graph.numEdgs = total;
    return graph;
}

/* ---------- BipartiteMatching construction ---------- */

BipartiteMatching emptyBipartiteMatching(const BipartiteGraph& graph) {
    BipartiteMatching matching;
    matching.sNumVtxs = graph.sNumVtxs;
    matching.tNumVtxs = graph.tNumVtxs;
    matching.numEdgs = 0;
    matching.sMate.assign(graph.sNumVtxs, NIL);
    matching.tMate.assign(graph.tNumVtxs, NIL);
    return matching;
}

/* ---------- Greedy initial matching: simple ---------- */

int32_t greedyInit(const BipartiteGraph& graph, BipartiteMatching& matching) {
    int32_t numEdgs = 0;
    for (size_t s = 0; s < graph.sNumVtxs; s++) {
        if (matching.sMate[s] != NIL) continue;
        for (int32_t t : graph.sAdj[s]) {
            if (matching.tMate[t] == NIL) {
                matching.sMate[s] = t;
                matching.tMate[t] = static_cast<int32_t>(s);
                numEdgs++;
                break;
            }
        }
    }
    matching.numEdgs += numEdgs;
    return numEdgs;
}

/* ---------- Greedy initial matching: min-degree ---------- */

int32_t greedyInitMd(const BipartiteGraph& graph, BipartiteMatching& matching) {
    int32_t numEdgs = 0;
    std::vector<int32_t> deg(graph.tNumVtxs, 0);
    for (size_t s = 0; s < graph.sNumVtxs; s++)
        for (int32_t t : graph.sAdj[s])
            deg[t]++;

    std::vector<int32_t> sOrder(graph.sNumVtxs);
    for (size_t s = 0; s < graph.sNumVtxs; s++) sOrder[s] = static_cast<int32_t>(s);
    /* Sort s-vertices in increasing order of degree, breaking ties by vertex label. */
    std::sort(sOrder.begin(), sOrder.end(), [&](int32_t s1, int32_t s2){
        size_t s1Deg = graph.sAdj[s1].size();
        size_t s2Deg = graph.sAdj[s2].size();
        return s1Deg < s2Deg || (s1Deg == s2Deg && s1 < s2);
    });

    for (int32_t s : sOrder) {
        if (matching.sMate[s] != NIL) continue;
        int32_t best = NIL, bestDeg = INT_MAX;
        for (int32_t t : graph.sAdj[s]) {
            if (matching.tMate[t] == NIL && deg[t] < bestDeg) {
                best = t;
                bestDeg = deg[t];
            }
        }
        if (best != NIL) {
            matching.sMate[s] = best;
            matching.tMate[best] = s;
            numEdgs++;
        }
    }
    matching.numEdgs += numEdgs;
    return numEdgs;
}

/* ---------- HK BFS ---------- */

static bool bfs(const BipartiteGraph& graph, const BipartiteMatching& matching, HKIState& state) {
    std::vector<int32_t> sPrcbQue(graph.sNumVtxs);
    int32_t queHead = 0, queTail = 0;

    for (size_t s = 0; s < graph.sNumVtxs; s++) {
        if (matching.sMate[s] == NIL) { state.sLevel[s] = 0; sPrcbQue[queTail++] = static_cast<int32_t>(s); }
        else state.sLevel[s] = INT_MAX;
    }
    state.sLevel[graph.sNumVtxs] = INT_MAX;  /* NIL sentinel */

    while (queHead < queTail) {
        int32_t s = sPrcbQue[queHead++];
        if (state.sLevel[s] < state.sLevel[graph.sNumVtxs]) {
            for (int32_t t : graph.sAdj[s]) {
                int32_t ss = (matching.tMate[t] == NIL) ? static_cast<int32_t>(graph.sNumVtxs) : matching.tMate[t];
                if (state.sLevel[ss] == INT_MAX) {
                    state.sLevel[ss] = state.sLevel[s] + 1;
                    if (matching.tMate[t] != NIL) sPrcbQue[queTail++] = matching.tMate[t];
                }
            }
        }
    }
    return state.sLevel[graph.sNumVtxs] != INT_MAX;
}

/*
 * DFS: iterative with edge index.
 *
 * state.sIdx[s] is the position within s's adjacency (graph.sAdj[s][state.sIdx[s]]),
 * persistent within a phase so dead-end edges are not revisited.
 */
static bool dfs(int32_t sFirst, const BipartiteGraph& graph, BipartiteMatching& matching, HKIState& state) {
    state.stkTop = 0;
    state.sPrcbStk[state.stkTop++] = sFirst;

    while (state.stkTop > 0) {
        int32_t s = state.sPrcbStk[state.stkTop - 1];
        int32_t sNumEdgs = static_cast<int32_t>(graph.sAdj[s].size());

        bool pushed = false;
        while (state.sIdx[s] < sNumEdgs) {
            int32_t t = graph.sAdj[s][state.sIdx[s]];
            int32_t ss = (matching.tMate[t] == NIL) ? static_cast<int32_t>(graph.sNumVtxs) : matching.tMate[t];
            if (state.sLevel[ss] != state.sLevel[s] + 1) {
                state.sIdx[s]++;
                continue;
            }

            state.tPrcbStk[state.stkTop - 1] = t;
            state.sIdx[s]++;

            if (matching.tMate[t] == NIL) {
                /* Found augmenting path — augment all the way back */
                for (int32_t k = state.stkTop - 1; k >= 0; k--) {
                    matching.tMate[state.tPrcbStk[k]] = state.sPrcbStk[k];
                    matching.sMate[state.sPrcbStk[k]] = state.tPrcbStk[k];
                }
                return true;
            }

            state.sPrcbStk[state.stkTop++] = matching.tMate[t];
            pushed = true;
            break;
        }

        if (!pushed) {
            state.sLevel[s] = INT_MAX;
            state.stkTop--;
        }
    }
    return false;
}

/* ---------- Top-level Hopcroft-Karp Iterative ---------- */

int32_t hkIterativeMcm(const BipartiteGraph& graph, BipartiteMatching& matching) {
    HKIState state;
    state.sLevel.assign(graph.sNumVtxs + 1, 0);
    state.sIdx.assign(graph.sNumVtxs, 0);
    state.sPrcbStk.assign(graph.sNumVtxs, 0);
    state.tPrcbStk.assign(graph.sNumVtxs, 0);
    state.stkTop = 0;

    int32_t numPhases = 0;
    int32_t newEdgs = 0;
    while (bfs(graph, matching, state)) {
        numPhases++;
        for (size_t s = 0; s < graph.sNumVtxs; s++) state.sIdx[s] = 0;
        for (size_t s = 0; s < graph.sNumVtxs; s++) {
            if (matching.sMate[s] == NIL && dfs(static_cast<int32_t>(s), graph, matching, state)) {
                newEdgs++;
            }
        }
    }
    matching.numEdgs += newEdgs;
    return numPhases;
}

/* ---------- Validation ---------- */

void validateBipartiteMatching(const BipartiteGraph& graph, const BipartiteMatching& matching) {
    int32_t errors = 0;
    int32_t matchedS = 0, matchedT = 0;

    for (size_t s = 0; s < graph.sNumVtxs; s++) {
        if (matching.sMate[s] != NIL) {
            matchedS++;
            int32_t t = matching.sMate[s];
            if (t < 0 || static_cast<size_t>(t) >= graph.tNumVtxs) {
                fprintf(stderr, "ERROR: sMate[%zu] = %d out of range\n", s, t);
                errors++;
            } else if (matching.tMate[t] != static_cast<int32_t>(s)) {
                fprintf(stderr, "ERROR: sMate[%zu]=%d but tMate[%d]=%d\n", s, t, t, matching.tMate[t]);
                errors++;
            } else if (!std::binary_search(graph.sAdj[s].begin(), graph.sAdj[s].end(), t)) {
                fprintf(stderr, "ERROR: edge (%zu,%d) not in graph\n", s, t);
                errors++;
            }
        }
    }
    for (size_t t = 0; t < graph.tNumVtxs; t++) {
        if (matching.tMate[t] != NIL) matchedT++;
    }

    printf("\n=== Validation Report ===\n");
    printf("Matching size: %zu\n", matching.numEdgs);
    printf("S matched: %d, T matched: %d\n", matchedS, matchedT);
    printf("%s\n", errors > 0 ? "VALIDATION FAILED" : "VALIDATION PASSED");
    printf("=========================\n\n");
}

/* ---------- Main ---------- */

int main(int argc, char* argv[]) {
    printf("Hopcroft-Karp Iterative Algorithm - C++ Implementation (VV)\n");
    printf("=============================================================\n\n");

    if (argc < 2) { printf("Usage: %s <filename> [--greedy|--greedy-md]\n", argv[0]); return 1; }
    int32_t greedyMode = 0;
    for (int i = 2; i < argc; i++) {
        if (std::string(argv[i]) == "--greedy") greedyMode = 1;
        else if (std::string(argv[i]) == "--greedy-md") greedyMode = 2;
    }

    FILE* f = fopen(argv[1], "r");
    if (!f) { fprintf(stderr, "Cannot open file: %s\n", argv[1]); return 1; }

    int sNumVtxs, tNumVtxs, numEdgsIn;
    if (fscanf(f, "%d %d %d", &sNumVtxs, &tNumVtxs, &numEdgsIn) != 3) {
        fprintf(stderr, "Bad header\n"); fclose(f); return 1;
    }

    std::vector<std::pair<int32_t,int32_t>> edges;
    edges.reserve(numEdgsIn);
    for (int i = 0; i < numEdgsIn; i++) {
        int s, t;
        if (fscanf(f, "%d %d", &s, &t) != 2) break;
        edges.push_back({static_cast<int32_t>(s), static_cast<int32_t>(t)});
    }
    fclose(f);

    printf("Graph: %d s-vertices, %d t-vertices, %zu edges\n",
           sNumVtxs, tNumVtxs, edges.size());

    BipartiteGraph graph = buildBipartiteGraph(sNumVtxs, tNumVtxs, edges);
    BipartiteMatching matching = emptyBipartiteMatching(graph);

    auto t0 = std::chrono::high_resolution_clock::now();

    int32_t greedySize = 0;
    if (greedyMode == 1) greedySize = greedyInit(graph, matching);
    else if (greedyMode == 2) greedySize = greedyInitMd(graph, matching);

    int32_t numPhases = hkIterativeMcm(graph, matching);

    auto t1 = std::chrono::high_resolution_clock::now();

    validateBipartiteMatching(graph, matching);

    printf("Phases: %d\n", numPhases);
    printf("Matching size: %zu\n", matching.numEdgs);
    if (greedyMode > 0) {
        printf("Greedy init size: %d\n", greedySize);
        if (matching.numEdgs > 0)
            printf("Greedy/Final: %.2f%%\n", 100.0 * greedySize / matching.numEdgs);
        else
            printf("Greedy/Final: NA\n");
    }
    printf("Time: %ld ms\n",
           static_cast<long>(std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count()));
    return 0;
}

File: algorithms/hk-it-vv/rs/hk_it_vv.rs
/*
 * Hopcroft-Karp Hybrid Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * VV (vector-of-vectors) adjacency.
 * Old HK's lean BFS (single dist[] array, sentinel trick) + iterative
 * stack-based DFS with edge index array (no recursion, no rescan).
 *
 * Refactored to separate input (BipartiteGraph), output (BipartiteMatching),
 * and algorithm state (HKHState).
 */

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

const NIL: i32 = -1;

/* ---------- Input: BipartiteGraph ---------- */

#[allow(dead_code)]
struct BipartiteGraph {
    s_num_vtxs: usize,
    t_num_vtxs: usize,
    num_edgs: usize,
    s_adj: Vec<Vec<i32>>,
    t_adj: Vec<Vec<i32>>,
}

/* ---------- Output: BipartiteMatching ---------- */

#[allow(dead_code)]
struct BipartiteMatching {
    s_num_vtxs: usize,
    t_num_vtxs: usize,
    num_edgs: usize,
    s_mate: Vec<i32>,
    t_mate: Vec<i32>,
}

/* ---------- State: HKHState ---------- */

struct HKHState {
    dist: Vec<i32>,         // length s_num_vtxs+1; dist[s_num_vtxs] is the NIL sentinel
    edge_idx: Vec<usize>,   // length s_num_vtxs; resume position per s-vertex within a phase
    stk_u: Vec<i32>,        // length s_num_vtxs; DFS stack (s-vertices)
    stk_v: Vec<i32>,        // length s_num_vtxs; t-vertex chosen at each depth
    stk_top: usize,         // current depth of stack
}

/* ---------- BipartiteGraph construction ---------- */

fn build_bipartite_graph(s_num_vtxs: usize, t_num_vtxs: usize,
                         edges: &[(usize, usize)]) -> BipartiteGraph {
    let mut s_adj: Vec<Vec<i32>> = vec![Vec::new(); s_num_vtxs];
    let mut t_adj: Vec<Vec<i32>> = vec![Vec::new(); t_num_vtxs];
    for &(u, v) in edges {
        if u < s_num_vtxs && v < t_num_vtxs {
            s_adj[u].push(v as i32);
            t_adj[v].push(u as i32);
        }
    }
    for adj in &mut s_adj { adj.sort_unstable(); adj.dedup(); }
    for adj in &mut t_adj { adj.sort_unstable(); adj.dedup(); }

    let num_edgs: usize = s_adj.iter().map(|a| a.len()).sum();

    BipartiteGraph {
        s_num_vtxs,
        t_num_vtxs,
        num_edgs,
        s_adj,
        t_adj,
    }
}

/* ---------- BipartiteMatching construction ---------- */

fn empty_bipartite_matching(g: &BipartiteGraph) -> BipartiteMatching {
    BipartiteMatching {
        s_num_vtxs: g.s_num_vtxs,
        t_num_vtxs: g.t_num_vtxs,
        num_edgs: 0,
        s_mate: vec![NIL; g.s_num_vtxs],
        t_mate: vec![NIL; g.t_num_vtxs],
    }
}

/* ---------- Greedy initial matching: simple ---------- */

fn greedy_init(g: &BipartiteGraph, m: &mut BipartiteMatching) -> usize {
    let mut cnt: usize = 0;
    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] != NIL { continue; }
        for &v in &g.s_adj[u] {
            if m.t_mate[v as usize] == NIL {
                m.s_mate[u] = v;
                m.t_mate[v as usize] = u as i32;
                cnt += 1;
                break;
            }
        }
    }
    m.num_edgs += cnt;
    cnt
}

/* ---------- Greedy initial matching: min-degree ---------- */

fn greedy_init_md(g: &BipartiteGraph, m: &mut BipartiteMatching) -> usize {
    let mut cnt: usize = 0;
    let mut deg = vec![0usize; g.t_num_vtxs];
    for u in 0..g.s_num_vtxs {
        for &v in &g.s_adj[u] {
            deg[v as usize] += 1;
        }
    }

    let mut order: Vec<usize> = (0..g.s_num_vtxs).collect();
    order.sort_unstable_by(|&a, &b| {
        g.s_adj[a].len().cmp(&g.s_adj[b].len()).then(a.cmp(&b))
    });

    for u in order {
        if m.s_mate[u] != NIL { continue; }
        let mut best: i32 = NIL;
        let mut best_deg = usize::MAX;
        for &v in &g.s_adj[u] {
            if m.t_mate[v as usize] == NIL && deg[v as usize] < best_deg {
                best = v;
                best_deg = deg[v as usize];
            }
        }
        if best != NIL {
            m.s_mate[u] = best;
            m.t_mate[best as usize] = u as i32;
            cnt += 1;
        }
    }
    m.num_edgs += cnt;
    cnt
}

/* ---------- HK BFS (same as basic HK) ---------- */

fn bfs(g: &BipartiteGraph, m: &BipartiteMatching, s: &mut HKHState) -> bool {
    let mut queue: Vec<usize> = Vec::with_capacity(g.s_num_vtxs);

    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] == NIL {
            s.dist[u] = 0;
            queue.push(u);
        } else {
            s.dist[u] = i32::MAX;
        }
    }
    s.dist[g.s_num_vtxs] = i32::MAX;

    let mut qi = 0;
    while qi < queue.len() {
        let u = queue[qi];
        qi += 1;
        if s.dist[u] < s.dist[g.s_num_vtxs] {
            for &v in &g.s_adj[u] {
                let paired = if m.t_mate[v as usize] == NIL {
                    g.s_num_vtxs
                } else {
                    m.t_mate[v as usize] as usize
                };
                if s.dist[paired] == i32::MAX {
                    s.dist[paired] = s.dist[u] + 1;
                    if m.t_mate[v as usize] != NIL {
                        queue.push(m.t_mate[v as usize] as usize);
                    }
                }
            }
        }
    }
    s.dist[g.s_num_vtxs] != i32::MAX
}

/*
 * DFS: iterative with edge index.
 *
 * The stack represents the alternating path being explored:
 *   stk_u[0] -> stk_v[0] -> stk_u[1] -> stk_v[1] -> ...
 *
 * edge_idx[u] persists across DFS calls within a phase, so edges
 * that led to dead ends are never revisited.
 */
fn dfs(root: usize, g: &BipartiteGraph, m: &mut BipartiteMatching, s: &mut HKHState) -> bool {
    s.stk_top = 0;
    s.stk_u[0] = root as i32;
    s.stk_top = 1;

    while s.stk_top > 0 {
        let u = s.stk_u[s.stk_top - 1] as usize;
        let sz = g.s_adj[u].len();

        let mut pushed = false;
        while s.edge_idx[u] < sz {
            let v = g.s_adj[u][s.edge_idx[u]];
            let paired = if m.t_mate[v as usize] == NIL {
                g.s_num_vtxs
            } else {
                m.t_mate[v as usize] as usize
            };
            if s.dist[paired] != s.dist[u] + 1 {
                s.edge_idx[u] += 1;
                continue;
            }

            /* Record which v this u chose */
            s.stk_v[s.stk_top - 1] = v;
            s.edge_idx[u] += 1;

            if m.t_mate[v as usize] == NIL {
                /* Found augmenting path — augment all the way back */
                let mut d = s.stk_top as isize - 1;
                while d >= 0 {
                    let du = s.stk_u[d as usize] as usize;
                    let dv = s.stk_v[d as usize] as usize;
                    m.t_mate[dv] = du as i32;
                    m.s_mate[du] = dv as i32;
                    d -= 1;
                }
                return true;
            }

            /* Go deeper: push t_mate[v] */
            s.stk_u[s.stk_top] = m.t_mate[v as usize];
            s.stk_top += 1;
            pushed = true;
            break;
        }

        if !pushed {
            /* Dead end — mark and backtrack */
            s.dist[u] = i32::MAX;
            s.stk_top -= 1;
        }
    }
    false
}

/* ---------- Top-level Hopcroft-Karp Hybrid ---------- */

fn hopcroft_karp_hybrid(g: &BipartiteGraph, m: &mut BipartiteMatching) {
    let n = g.s_num_vtxs.max(1);
    let mut s = HKHState {
        dist: vec![0i32; g.s_num_vtxs + 1],
        edge_idx: vec![0usize; g.s_num_vtxs],
        stk_u: vec![0i32; n],
        stk_v: vec![0i32; n],
        stk_top: 0,
    };

    let mut phases: i32 = 0;
    let mut new_edgs: usize = 0;
    while bfs(g, m, &mut s) {
        phases += 1;
        /* Reset edge indices for this phase */
        for u in 0..g.s_num_vtxs { s.edge_idx[u] = 0; }
        for u in 0..g.s_num_vtxs {
            if m.s_mate[u] == NIL && dfs(u, g, m, &mut s) {
                new_edgs += 1;
            }
        }
    }
    m.num_edgs += new_edgs;
    println!("Phases: {}", phases);
}

/* ---------- Validation ---------- */

fn validate_bipartite_matching(g: &BipartiteGraph, m: &BipartiteMatching) {
    let mut errors = 0;
    let mut matched_s = 0;
    let mut matched_t = 0;

    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] != NIL {
            matched_s += 1;
            let v = m.s_mate[u];
            if v < 0 || (v as usize) >= g.t_num_vtxs {
                eprintln!("ERROR: s_mate[{}] = {} out of range", u, v);
                errors += 1;
            } else if m.t_mate[v as usize] != u as i32 {
                eprintln!("ERROR: s_mate[{}]={} but t_mate[{}]={}", u, v, v, m.t_mate[v as usize]);
                errors += 1;
            } else if !g.s_adj[u].binary_search(&v).is_ok() {
                eprintln!("ERROR: edge ({},{}) not in graph", u, v);
                errors += 1;
            }
        }
    }
    for v in 0..g.t_num_vtxs {
        if m.t_mate[v] != NIL { matched_t += 1; }
    }

    println!("\n=== Validation Report ===");
    println!("Matching size: {}", m.num_edgs);
    println!("S matched: {}, T matched: {}", matched_s, matched_t);
    println!("{}", if errors > 0 { "VALIDATION FAILED" } else { "VALIDATION PASSED" });
    println!("=========================\n");
}

/* ---------- Graph loader ---------- */

fn load_graph(filename: &str) -> Result<(usize, usize, Vec<(usize, usize)>), Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    let first = lines.next().ok_or("Empty file")??;
    let parts: Vec<&str> = first.split_whitespace().collect();
    if parts.len() != 3 {
        return Err("First line must have 3 numbers".into());
    }
    let s_num_vtxs: usize = parts[0].parse()?;
    let t_num_vtxs: usize = parts[1].parse()?;
    let m: usize = parts[2].parse()?;

    let mut edges = Vec::with_capacity(m);
    for line in lines {
        let line = line?;
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            let u: usize = parts[0].parse()?;
            let v: usize = parts[1].parse()?;
            edges.push((u, v));
        }
    }
    Ok((s_num_vtxs, t_num_vtxs, edges))
}

/* ---------- Main ---------- */

fn main() {
    println!("Hopcroft-Karp Hybrid Algorithm - Rust Implementation (VV)");
    println!("===========================================================\n");

    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <filename> [--greedy|--greedy-md]", args[0]);
        std::process::exit(1);
    }
    let greedy_mode: i32 = if args.iter().any(|a| a == "--greedy-md") { 2 }
                          else if args.iter().any(|a| a == "--greedy") { 1 }
                          else { 0 };

    match load_graph(&args[1]) {
        Ok((s_num_vtxs, t_num_vtxs, edges)) => {
            println!("Graph: {} s-vertices, {} t-vertices, {} edges",
                     s_num_vtxs, t_num_vtxs, edges.len());

            let bipartite_graph = build_bipartite_graph(s_num_vtxs, t_num_vtxs, &edges);
            let mut bipartite_matching = empty_bipartite_matching(&bipartite_graph);

            let start = Instant::now();

            let greedy_size: usize = match greedy_mode {
                1 => greedy_init(&bipartite_graph, &mut bipartite_matching),
                2 => greedy_init_md(&bipartite_graph, &mut bipartite_matching),
                _ => 0,
            };

            hopcroft_karp_hybrid(&bipartite_graph, &mut bipartite_matching);

            let duration = start.elapsed();

            validate_bipartite_matching(&bipartite_graph, &bipartite_matching);

            println!("Matching size: {}", bipartite_matching.num_edgs);
            if greedy_mode > 0 {
                let fs = bipartite_matching.num_edgs;
                println!("Greedy init size: {}", greedy_size);
                if fs > 0 {
                    println!("Greedy/Final: {:.2}%", 100.0 * greedy_size as f64 / fs as f64);
                } else {
                    println!("Greedy/Final: NA");
                }
            }
            println!("Time: {} ms", duration.as_millis());
        }
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    }
}
